README.md
# Go bindings for MusicBrainz libdiscid
[![PkgGoDev](https://pkg.go.dev/badge/go.uploadedlobster.com/discid)](https://pkg.go.dev/go.uploadedlobster.com/discid)

**This project has been moved to https://git.sr.ht/~phw/go-discid**

## About
discid provides Go bindings for the MusicBrainz DiscID library [libdiscid](http://musicbrainz.org/doc/libdiscid).
It allows calculating DiscIDs (MusicBrainz and freedb) for Audio CDs. Additionally
the library can extract the MCN/UPC/EAN and the ISRCs from disc.

## Requirements
* libdiscid >= 0.6.0

## Usage

```go
package main

import (
	"fmt"
	"log"

	"go.uploadedlobster.com/discid"
)

func main() {
  // Specifying the device is optional. If set to an empty string a platform
  // specific default will be used.
  disc, err := discid.ReadFeatures("", discid.FeatureRead|discid.FeatureMcn)
  if err != nil {
    log.Fatal(err)
  }
  defer disc.Close()
  fmt.Printf("Disc ID: %v\n", disc.Id())
  fmt.Printf("MCN    : %v\n", disc.Mcn())
}
```

See the [API documentation](https://pkg.go.dev/go.uploadedlobster.com/discid) for details.

## Contribute
The source code for discid-sys is available on
[SourceHut](https://git.sr.ht/~phw/go-discid).

Please report any issues on the
[issue tracker](https://todo.sr.ht/~phw/discid-bindings).

Patches can be submitted to the [mailing list](https://lists.sr.ht/~phw/musicbrainz).
You can clone the repository directly on SourceHut and submit your changes
with the "Prepare patchset" button. Please see SourceHut's
[documentation for sending patches upstream](https://man.sr.ht/git.sr.ht/#sending-patches-upstream)
for details.

## License
discid Copyright (c) 2020-2023 by Philipp Wolfer <ph.wolfer@gmail.com>

discid is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

See [LICENSE](./LICENSE) for details.


CHANGES.md
# Changelog

## 0.3.0 (2023-02-28)
- Changed module path to `go.uploadedlobster.com/discid`
- Update testify to v1.8.2

## 0.2.1 (2022-07-19)
Update testify dependency

## 0.2.0 (2021-09-24)
Added examples `discinfo` and `discinfo-cgo`

## 0.1.0 (2020-11-11)
Initial release


discid_test.go
// Copyright (C) 2020-2023 Philipp Wolfer <ph.wolfer@gmail.com>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

package discid_test

import (
	"fmt"
	"log"
	"strconv"
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
	"go.uploadedlobster.com/discid"
)

func TestDefaultDevice(t *testing.T) {
	device := discid.DefaultDevice()
	if device == "" {
		t.Errorf("TestDefaultDevice() is empty; expected device name")
	}
}

func ExampleDefaultDevice() {
	fmt.Printf("Default device: %v\n", discid.DefaultDevice())
}

func TestVersion(t *testing.T) {
	version := discid.Version()
	if !strings.HasPrefix(version, "libdiscid") {
		t.Errorf("Version() = %v; expected starting with \"libdiscid\"", version)
	}
}

func ExampleVersion() {
	fmt.Printf("Version: %v\n", discid.Version())
}

func TestHasFeature(t *testing.T) {
	result := discid.HasFeature(discid.FeatureRead)
	if !result {
		t.Errorf("HasFeature() = %v; expected true", result)
	}
}

func ExampleHasFeature() {
	if discid.HasFeature(discid.FeatureIsrc) {
		fmt.Println("ISRC support available")
	}
}

func TestReadInvalidDevice(t *testing.T) {
	_, err := discid.Read("notadevice")
	if err == nil {
		t.Errorf("Expected error for accessing invalid device")
	}
}

func ExampleRead() {
	disc, err := discid.Read("") // Read from default device
	if err != nil {
		log.Fatal(err)
	}
	defer disc.Close()
	fmt.Printf("Disc ID: %v\n", disc.Id())
}

func ExampleReadFeatures() {
	// Read TOC and MCN from the disc in /dev/cdrom
	disc, err := discid.ReadFeatures("/dev/cdrom", discid.FeatureRead|discid.FeatureMcn)
	if err != nil {
		log.Fatal(err)
	}
	defer disc.Close()
	fmt.Printf("Disc ID: %v\n", disc.Id())
	fmt.Printf("MCN    : %v\n", disc.Mcn())
}

func TestPut(t *testing.T) {
	assert := assert.New(t)
	first := 1
	offsets := []int{
		206535, 150, 18901, 39738, 59557, 79152, 100126, 124833, 147278, 166336, 182560,
	}
	disc, err := discid.Put(first, offsets)
	if err != nil {
		t.Fatal(err)
	}
	defer disc.Close()
	assert.Equal("Wn8eRBtfLDfM0qjYPdxrz.Zjs_U-", disc.Id())
	assert.Equal(disc.Id(), fmt.Sprint(disc))
	assert.Equal("830abf0a", disc.FreedbId())
	assert.Equal(1, disc.FirstTrackNum())
	assert.Equal(10, disc.LastTrackNum())
	assert.Equal(206535, disc.Sectors())
	assert.Equal(
		"1 10 206535 150 18901 39738 59557 79152 100126 124833 147278 166336 182560",
		disc.TocString())
	assert.Equal(
		"http://musicbrainz.org/cdtoc/attach?id=Wn8eRBtfLDfM0qjYPdxrz.Zjs_U-&tracks=10&toc=1+10+206535+150+18901+39738+59557+79152+100126+124833+147278+166336+182560",
		disc.SubmissionUrl())
	for i := disc.FirstTrackNum(); i <= disc.LastTrackNum(); i++ {
		track := disc.Track(i)
		offset := offsets[track.Number]
		next := 0
		if track.Number < disc.LastTrackNum() {
			next = track.Number + 1
		}
		length := offsets[next] - offset
		assert.Equal(i, track.Number)
		assert.Equal(offset, track.Offset)
		assert.Equal(length, track.Sectors)
		assert.Equal("", track.Isrc)
	}
}

func TestPutFirstTrackLargerOne(t *testing.T) {
	assert := assert.New(t)
	first := 3
	offsets := []int{
		206535, 150, 18901, 39738, 59557, 79152, 100126, 124833, 147278, 166336, 182560,
	}
	disc, err := discid.Put(first, offsets[0:])
	if err != nil {
		t.Fatal(err)
	}
	defer disc.Close()
	assert.Equal("ByBKvJM1hBL7XtvsPyYtIjlX0Bw-", disc.Id())
	assert.Equal(3, disc.FirstTrackNum())
	assert.Equal(12, disc.LastTrackNum())
	assert.Equal(206535, disc.Sectors())
}

func TestPutTooManyOffsets(t *testing.T) {
	first := 1
	offsets := [101]int{}
	disc, err := discid.Put(first, offsets[0:])
	assert.Empty(t, disc)
	assert.NotEmpty(t, err)
	if err.Error() != "Illegal track limits" {
		t.Errorf("Expected error \"Illegal track limits\"")
	}
}

func TestPutTooManyTracks(t *testing.T) {
	// First track number is 82, 19 tracks
	// => last track number would be 100, but 99 is max.
	first := 82
	offsets := [20]int{}
	disc, err := discid.Put(first, offsets[0:])
	assert.Empty(t, disc)
	assert.NotEmpty(t, err)
	if err.Error() != "Illegal track limits" {
		t.Errorf("Expected error \"Illegal track limits\"")
	}
}

func ExamplePut() {
	first := 1
	offsets := []int{
		242457, 150, 44942, 61305, 72755, 96360, 130485, 147315, 164275, 190702, 205412, 220437,
	}
	disc, err := discid.Put(first, offsets)
	if err != nil {
		log.Fatal(err)
	}
	defer disc.Close()
	fmt.Println(disc.Id())
	// Output: lSOVc5h6IXSuzcamJS1Gp4_tRuA-
}

func ExampleParse() {
	toc := "1 11 242457 150 44942 61305 72755 96360 130485 147315 164275 190702 205412 220437"
	disc, err := discid.Parse(toc)
	if err != nil {
		log.Fatal(err)
	}
	defer disc.Close()
	fmt.Println(disc.Id())
	// Output: lSOVc5h6IXSuzcamJS1Gp4_tRuA-
}

func TestParseMinimal(t *testing.T) {
	assert := assert.New(t)
	toc := "1 1 44942 150"
	disc, err := discid.Parse(toc)
	if err != nil {
		t.Fatal(err)
	}
	defer disc.Close()
	assert.Equal("ANJa4DGYN_ktpzOwvVPtcjwP7mE-", disc.Id())
	assert.Equal(toc, disc.TocString())
}

func TestParseFirstTrackNotOne(t *testing.T) {
	assert := assert.New(t)
	toc := "3 12 242457 150 18901 39738 59557 79152 100126 124833 147278 166336 182560"
	disc, err := discid.Parse(toc)
	if err != nil {
		t.Fatal(err)
	}
	defer disc.Close()
	assert.Equal("fC1yNbC5bVjbvphqlAY9JyYoWEY-", disc.Id())
	assert.Equal(toc, disc.TocString())
}

func TestParseNaN(t *testing.T) {
	toc := "1 2 242457 150 a"
	_, err := discid.Parse(toc)
	if assert.Error(t, err) {
		if err.(*strconv.NumError).Err != strconv.ErrSyntax {
			t.Errorf("Expected strconv.ErrSyntax, got \"%v\"", err)
		}
	}
}

func TestParseInvalidEmpty(t *testing.T) {
	toc := ""
	_, err := discid.Parse(toc)
	if assert.Error(t, err) {
		if err.(*strconv.NumError).Err != strconv.ErrSyntax {
			t.Errorf("Expected strconv.ErrSyntax, got \"%v\"", err)
		}
	}
}

func TestParseTooManyOffsets(t *testing.T) {
	assert := assert.New(t)
	toc := "1 2 242457 150 200 300"
	_, err := discid.Parse(toc)
	assert.Error(err)
	assert.Equal("TOC string contains too many offsets (max. 100)", err.Error())
}

func TestParseTooManyOffsetsTotal(t *testing.T) {
	assert := assert.New(t)
	indexes := [103]string{"1", "99", "20000"}
	for i := 3; i < len(indexes); i++ {
		indexes[i] = strconv.Itoa(i * 100)
	}
	toc := strings.Join(indexes[:], " ")
	_, err := discid.Parse(toc)
	assert.Error(err)
	assert.Equal("TOC string contains too many offsets (max. 100)", err.Error())
}

func TestParseInvalidMissingOffsets(t *testing.T) {
	assert := assert.New(t)
	toc := "1 2 242457 150"
	_, err := discid.Parse(toc)
	assert.Error(err)
	assert.Equal("Number of offsets 1 does not match track count 2", err.Error())
}

func TestParseInvalidNotEnoughElements(t *testing.T) {
	assert := assert.New(t)
	toc := "1"
	_, err := discid.Parse(toc)
	assert.Error(err)
	assert.Equal("Invalid TOC string \"1\"", err.Error())
}

func TestTrackOutOfRange(t *testing.T) {
	assert := assert.New(t)
	first := 1
	offsets := []int{
		206535, 150, 18901, 39738, 59557, 79152, 100126, 124833, 147278, 166336, 182560,
	}
	disc, err := discid.Put(first, offsets)
	if err != nil {
		t.Fatal(err)
	}
	defer disc.Close()
	assert.Panics(func() { disc.Track(disc.FirstTrackNum() - 1) })
	assert.NotPanics(func() { disc.Track(disc.FirstTrackNum()) })
	assert.NotPanics(func() { disc.Track(disc.LastTrackNum()) })
	assert.Panics(func() { disc.Track(disc.LastTrackNum() + 1) })
}


examples/discinfo/main.go
// This example uses go-discid to query details about the disc in the default device.

package main

import (
	"fmt"
	"log"

	"go.uploadedlobster.com/discid"
)

func main() {
	fmt.Printf("Version       : %v\n", discid.Version())
	fmt.Printf("Default device: %v\n", discid.DefaultDevice())
	// Read from default device
	disc, err := discid.ReadFeatures("", discid.FeatureAll)
	if err != nil {
		log.Fatal(err)
	}
	defer disc.Close()
	fmt.Printf("Disc ID       : %v\n", disc.Id())
	fmt.Printf("FreeDB ID     : %v\n", disc.FreedbId())
	fmt.Printf("TOC           : %v\n", disc.TocString())
	fmt.Printf("MCN           : %v\n", disc.Mcn())
	fmt.Printf("First track   : %v\n", disc.FirstTrackNum())
	fmt.Printf("Last track    : %v\n", disc.LastTrackNum())
	fmt.Printf("Sectors       : %v\n\n", disc.Sectors())

	for n := disc.FirstTrackNum(); n <= disc.LastTrackNum(); n++ {
		track := disc.Track(n)
		fmt.Printf("Track #%v:\n", track.Number)
		fmt.Printf("    ISRC   : %v\n", track.Isrc)
		fmt.Printf("    Offset : %v\n", track.Offset)
		fmt.Printf("    Sectors: %v\n", track.Sectors)
	}
}


discid.go
// Copyright (C) 2020-2023 Philipp Wolfer <ph.wolfer@gmail.com>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

// discid allows you to calculate MusicBrainz or FreeDB disc IDs for audio CDs.
//
// discid provides bindings to the MusicBrainz libdiscid (https://musicbrainz.org/doc/libdiscid)
// library. In addition to calculating the disc IDs you can also get advanced data from the
// audio CD such as MCN (media catalogue number) and per-track ISRCs.
//
// To get started see the documentation and examples of discid.Read, discid.ReadFeatures,
// discid.Put and discid.Parse.
//
// Details about the use and calculation of disc IDs can be found at the MusicBrainz
// disc ID documentation (https://musicbrainz.org/doc/Disc_ID).
//
// The source code of this library is available on SourceHut (https://git.sr.ht/~phw/go-discid)
// under the terms of the GNU Lesser General Public License version 3 or later.
package discid

// #cgo LDFLAGS: -ldiscid
// #include <stdlib.h>
// #include "discid/discid.h"
import "C"
import (
	"errors"
	"fmt"
	"strconv"
	"strings"
	"unsafe"
)

// Platform dependent feature
//
// The platform dependent features are currently discid.FeatureRead,
// discid.FeatureMcn and discid.FeatureIsrc.
//
// See the libdiscid feature matrix (https://musicbrainz.org/doc/libdiscid#Feature_Matrix)
// for a list of supported features per platform.
type Feature uint

const (
	// Read TOC from disc
	FeatureRead = C.DISCID_FEATURE_READ
	// Read MCN from disc
	FeatureMcn = C.DISCID_FEATURE_MCN
	// Read ISRCs from disc
	FeatureIsrc = C.DISCID_FEATURE_ISRC
	// Read with all features
	FeatureAll = FeatureRead | FeatureMcn | FeatureIsrc
)

// Holds information about a read disc (TOC, MCN, ISRCs).
//
// Use discid.Read, discid.ReadFeatures, discid.Put or discid.Parse
// to initialize an instance of Disc.
//
// Use the Close method to free the allocated resources after use, e.g.:
//
//	disc := discid.Read("") // Read from default device
//	defer disc.Close()
type Disc struct {
	handle *C.DiscId
}

// Holds information about a single track
type Track struct {
	// Track number (1-99) of the track
	Number int
	// Start offset in sectors
	Offset int
	// Track length in sectors
	Sectors int
	// ISRC for this track (might be empty).
	//
	// This will only bet set if discid.ReadFeatures` is called with discid.FeatureIsrc.
	Isrc string
}

// Return the name of the default disc drive for this operating system.
//
// The default device is system dependent, e.g. "/dev/cdrom" on Linux and "D:" on Windows.
func DefaultDevice() string {
	device := C.discid_get_default_device()
	return C.GoString(device)
}

// Return version information about libdiscid.
//
// The returned string will be e.g. "libdiscid 0.6.2".
func Version() string {
	version := C.discid_get_version_string()
	return C.GoString(version)
}

// Check if a certain feature is implemented on the current platform.
//
// This only works for single features, not bit masks with multiple features.
//
// See the libdiscid feature matrix (https://musicbrainz.org/doc/libdiscid#Feature_Matrix)
// for a list of supported features per platform.
func HasFeature(feature Feature) bool {
	result := C.discid_has_feature(uint32(feature))
	return result == 1
}

// Read the disc in the given CD-ROM/DVD-ROM drive extracting only the TOC.
//
// This function reads the disc in the drive specified by the given device
// identifier. If the device is an empty string, the default device, as
// returned by discid.DefaultDevice, is used.
//
// This function will only read the TOC, hence only the disc ID itself will be
// available. Use discid::ReadFeatures if you want to read also MCN and ISRCs.
func Read(device string) (disc Disc, err error) {
	return ReadFeatures(device, FeatureRead)
}

// Read the disc in the given CD-ROM/DVD-ROM drive with additional features.
//
// This function is similar to disc.Read but allows to read information about
// MCN and per-track ISRCs in addition to the normal TOC data.
//
// The parameter features accepts a bitwise combination of values.
// discid.FeatureRead is always implied, so it is not necessary to specify it.
//
// Reading MCN and ISRCs is not available on all platforms. You can use the
// has_feature function to check if a specific feature is available. Passing
// unsupported features here will just be ignored.
//
// Note that reading MCN and ISRC data is significantly slower than just
// reading the TOC, so only request the features you actually need.
func ReadFeatures(device string, features Feature) (disc Disc, err error) {
	d := Disc{C.discid_new()}
	var c_device *C.char = nil
	if device != "" {
		c_device = C.CString(device)
		defer C.free(unsafe.Pointer(c_device))
	}
	var status = C.discid_read_sparse(d.handle, c_device, C.uint(features))
	if status == 0 {
		defer d.Close()
		err = errors.New(d.ErrorMessage())
	} else {
		disc = d
	}
	return
}

// Provides the TOC of a known CD.
//
// This function may be used if the TOC has been read earlier and you want to calculate
// the disc ID afterwards, without accessing the disc drive.
//
// first is the track number of the first track (1-99).
// The offsets parameter is an array which contains the track offsets for each track.
// The first element, offsets[0], is the leadout track. It must contain the total number of
// sectors on the disc. offsets must not be longer than 100 elements (leadout + 99 tracks).
func Put(first int, offsets []int) (disc Disc, err error) {
	last := first + len(offsets) - 2
	d := Disc{C.discid_new()}
	// libdiscid always expects an array of 100 integers, no matter the track count.
	var c_offsets [100]C.int
	c_offsets[0] = C.int(offsets[0])
	for i, n := range offsets[1:] {
		track := i + first
		if track > 99 {
			break
		}
		c_offsets[track] = C.int(n)
	}
	var status = C.discid_put(d.handle, C.int(first), C.int(last), &c_offsets[0])
	if status == 0 {
		defer d.Close()
		err = errors.New(d.ErrorMessage())
	} else {
		disc = d
	}
	return
}

// Parses a TOC string and returns a Disc instance for it.
//
// The TOC string provided here must have the same format as returned by Disc.TocString.
//
// This function can be used if you already have a TOC string like e.g.
// "1 11 242457 150 44942 61305 72755 96360 130485 147315 164275 190702 205412 220437".
func Parse(toc string) (disc Disc, err error) {
	first := 0
	last := 0
	var offsets [100]int
	var i int
	var part string
	for i, part = range strings.Split(toc, " ") {
		parsedInt, e := strconv.Atoi(part)
		if e != nil {
			err = e
			return
		}
		if i == 0 {
			first = parsedInt
		} else if i == 1 {
			last = parsedInt
		} else {
			if i > (last+2) || i > 99+2 {
				err = errors.New("TOC string contains too many offsets (max. 100)")
				return
			}
			offsets[i-2] = parsedInt
		}
	}

	if i < 2 || first < 1 || last < 1 || last > 99 {
		msg := fmt.Sprintf("Invalid TOC string \"%v\"", toc)
		err = errors.New(msg)
		return
	}

	offsetCount := i - 2
	trackCount := last - first + 1
	if offsetCount < trackCount {
		msg := fmt.Sprintf("Number of offsets %v does not match track count %v",
			offsetCount, trackCount)
		err = errors.New(msg)
		return
	}

	return Put(first, offsets[0:trackCount+1])
}

// Release the memory allocated for the Disc object.
func (d Disc) Close() {
	C.discid_free(d.handle)
}

// Return a human-readable error message.
//
// This function may only be used if discid.Read failed.
func (d Disc) ErrorMessage() string {
	err := C.discid_get_error_msg(d.handle)
	return C.GoString(err)
}

// String representation of the disc, same as Id()
func (d Disc) String() string {
	return d.Id()
}

// Returns the MusicBrainz disc ID.
func (d Disc) Id() string {
	id := C.discid_get_id(d.handle)
	return C.GoString(id)
}

// Returns the FreeDB disc ID.
func (d Disc) FreedbId() string {
	id := C.discid_get_freedb_id(d.handle)
	return C.GoString(id)
}

// Return a string representing CD Table Of Contents (TOC).
//
// The TOC string is a list of integers separated by a single space character.
//
// Example: 1 11 242457 150 44942 61305 72755 96360 130485 147315 164275 190702 205412 220437
//
// The integers represent (in order):
//
// - First track number (normally one)
//
// - Last track number
//
// - Lead-out track offset
//
// - Up to 99 frame offsets
func (d Disc) TocString() string {
	toc := C.discid_get_toc_string(d.handle)
	return C.GoString(toc)
}

// An URL for submitting the DiscID to MusicBrainz.
func (d Disc) SubmissionUrl() string {
	url := C.discid_get_submission_url(d.handle)
	return C.GoString(url)
}

// The number of the first track on this disc.
func (d Disc) FirstTrackNum() int {
	return int(C.discid_get_first_track_num(d.handle))
}

// The number of the last track on this disc.
func (d Disc) LastTrackNum() int {
	return int(C.discid_get_last_track_num(d.handle))
}

// The length of the disc in sectors.
func (d Disc) Sectors() int {
	return int(C.discid_get_sectors(d.handle))
}

// Return the Media Catalogue Number (MCN) for the disc, if present.
//
// This is essentially an EAN (= UPC with 0 prefix).
func (d Disc) Mcn() string {
	mcn := C.discid_get_mcn(d.handle)
	return C.GoString(mcn)
}

// Return the Media Catalogue Number (MCN) for the disc, if present.
//
// This is essentially an EAN (= UPC with 0 prefix).
func (d Disc) Track(number int) Track {
	first := d.FirstTrackNum()
	last := d.LastTrackNum()
	if number < first || number > last {
		err := fmt.Sprintf(
			"track number out of bounds: given %v, expected between %v and %v",
			number, first, last)
		panic(err)
	}
	n := C.int(number)
	return Track{
		number,
		int(C.discid_get_track_offset(d.handle, n)),
		int(C.discid_get_track_length(d.handle, n)),
		C.GoString(C.discid_get_track_isrc(d.handle, n)),
	}
}
